More safety changes for "Effective Dart". (#2986)
dart-lang/site-www: src/_guides/language/effective-dart/usage_migrated.md

}
{% endprettify %}

> One common reason that fields are *not* initialized at the declaration is
> because they need to reference `this` or call methods on the containing
> instance. However, if the field is marked `late`, then the initializer *can*
> access `this`.

Reviewer: This can be read as if fields always need access to `this`. How about something like this (which I think you could improve):

```suggestion
Some fields can't be initialized at declaration because they need to reference
`this` — to use other fields or call methods, for example. However, if the
field is marked `late`, then the initializer *can* access `this`.
```

}
{% endprettify %}

One common reason that fields are *not* initialized at the declaration is
because they need to reference `this` or call methods on the containing
instance. However, if the field is marked `late`, then the initializer *can*
access `this`. In cases where the initialization doesn't depend on constructor
parameters, you may be able to move the initialization to the field's
declaration. That doesn't necessarily mean you *should* use `late` just to be
able to initialize at the declaration. But if you want the field to be `late`,

> it gives you additional flexibility.

Reviewer 2: I don't think I understand this.

Why else would someone want the field to be late?

Is an author going to read this an push back against a CR comment saying "make this field late final so we can move the initialization out of the constructor" with "You should only do that if you want the field to be `late` and I don't have another reason to want that"?

Author: My intent was to not come across as strongly encouraging users to use `late` unless they actually need it. But I guess you're probably right that in cases where this applies, they probably *do* want to use `late` so it's fine. Removed those sentences.

initializing. But when you *can* use initializing formals, you *should*.

### DON'T use `late` when a constructor initialization list will do.

> Sound null safety requires Dart to ensure that a non-nullable field must be

Reviewer: "ensure" and "must be" seem redundant. Maybe:

```suggestion
Sound null safety requires Dart to ensure that a non-nullable field is
```

initialized before it can be read. Since fields can be read inside the
constructor body, this means you get an error if you don't initialize a
non-nullable field before the body runs.

You can make this error go away by marking the field `late`. That turns the
compile-time error into a *runtime* error if you access the field before it is
initialized. That's what you need in some cases, but often the right fix is to
initialize the field in the constructor initialization list:

{:.good}
<?code-excerpt "usage_good.dart (late-init-list)"?>
{% prettify dart tag=pre+code %}
class Point {
  double x, y;
  Point.polar(double theta, double radius)
      : x = cos(theta) * radius,
        y = sin(theta) * radius;
}
{% endprettify %}

{:.bad}
<?code-excerpt "usage_bad.dart (late-init-list)"?>
{% prettify dart tag=pre+code %}
class Point {
  late double x, y;
  Point.polar(double theta, double radius) {
    x = cos(theta) * radius;
    y = sin(theta) * radius;
  }
}
{% endprettify %}

The initialization list gives you access to constructor parameters and lets you

> initialize fields before they can be read. So, if it's possible to use that,

Reviewer: ```suggestion
initialize fields before they can be read. So, if it's possible to use an initialization list,
```

(There was quite a distance between "that" and its antecedent.)

Also... I just noticed that you use _initialization list_ but the language tour uses _initializer list_. I just checked the language spec, and it uses _initializer list_. Could you make that change throughout Effective Dart?

Author: Will do.

### DON'T use `late` when a constructor initialization list will do.

Sound null safety requires Dart to ensure that a non-nullable field must be
initialized before it can be read. Since fields can be read inside the
constructor body, this means you get an error if you don't initialize a
non-nullable field before the body runs.

You can make this error go away by marking the field `late`. That turns the
compile-time error into a *runtime* error if you access the field before it is
initialized. That's what you need in some cases, but often the right fix is to
initialize the field in the constructor initialization list:

{:.good}
<?code-excerpt "usage_good.dart (late-init-list)"?>
{% prettify dart tag=pre+code %}
class Point {
  double x, y;
  Point.polar(double theta, double radius)
      : x = cos(theta) * radius,
        y = sin(theta) * radius;
}
{% endprettify %}

{:.bad}
<?code-excerpt "usage_bad.dart (late-init-list)"?>
{% prettify dart tag=pre+code %}
class Point {
  late double x, y;
  Point.polar(double theta, double radius) {
    x = cos(theta) * radius;
    y = sin(theta) * radius;
  }
}
{% endprettify %}

> The initialization list gives you access to constructor parameters and lets you initialize fields before they can be read. So, if it's possible to use that, that's better than making the field `late` and losing some static safety and performance.

Reviewer 2: [nit] 80 columns?
